mempool/src/shared_mempool/priority.rs

```diff
@@ -208,11 +242,160 @@ impl PrioritizedPeersState {
             .collect()
     }
 
+    fn update_sender_bucket_for_peers(
+        &mut self,
+        peer_monitoring_data: &HashMap<PeerNetworkId, Option<&PeerMonitoringMetadata>>,
+        num_txns_received_since_peers_updated: u64,
+    ) {
+        // TODO: If the top peer set didn't change, then don't change the Primary sender bucket assignment.
+        // TODO: (Minor) If the load is low, don't do load balancing for Failover buckets.
+        assert!(self.prioritized_peers.read().len() == peer_monitoring_data.len());
+
+        // Obtain the top peers to assign the sender buckets with Primary priority
+        let mut top_peers = vec![];
+        let time_elapsed_since_last_update =
+            self.last_peer_priority_update.map_or(0, |last_update| {
+                self.time_service
+                    .now()
+                    .duration_since(last_update)
+                    .as_secs()
+            });
+
+        let average_mempool_traffic_observed = num_txns_received_since_peers_updated as f64
+            / max(1, time_elapsed_since_last_update) as f64;
+        // Obtain the highest threshold from mempool_config.load_balancing_thresholds for which avg_mempool_traffic_threshold_in_tps exceeds average_mempool_traffic_observed
+        let threshold_config = self
+            .mempool_config
+            .load_balancing_thresholds
+            .clone()
+            .into_iter()
+            .rev()
+            .find(|threshold_config| {
+                threshold_config.avg_mempool_traffic_threshold_in_tps
+                    <= average_mempool_traffic_observed as u64
+            })
+            .unwrap_or_default();
+
+        let num_top_peers = max(
+            1,
+            min(
+                self.mempool_config.num_sender_buckets,
+                if self.mempool_config.enable_max_load_balancing_at_any_load {
+                    u8::MAX
+                } else {
+                    threshold_config.max_number_of_upstream_peers
+                },
+            ),
+        );
+        info!(
+            "Time elapsed since last peer update: {:?}\n
+            Number of transactions received since last peer update: {:?},\n
+            Average mempool traffic observed: {:?},\n
+            Load balancing threshold config: {:?},\n
+            Number of top peers picked: {:?}",
+            time_elapsed_since_last_update,
+            num_txns_received_since_peers_updated,
+            average_mempool_traffic_observed,
+            threshold_config,
+            num_top_peers
+        );
+
+        if self.node_type.is_validator_fullnode() {
```

Comment:
@JoshLind are you comfortable with NodeType being used to special case this? It seems like NodeType was originally meant to be config-internal so wanted to check

It seems somewhat redundant because PFNs shouldn't have any peers with `NetworkId::Vfn` to begin with. But maybe being explicit is good.

Reply:
Yeah, feels kinda iffy to me. But, not the end of the world. At some point we need to clean up all of the node role/type structs up.

One thing I'm not sure about (I haven't looked at the entire PR), but if this code is only running on VFNs and PFNs, then the VFN peers should only be a single validator. The code below looks okay, but something to call out.

Reply:
> If this is only being run on PFNs, we won't have VFN peers.

This code is run on PFNs and VFNs. For PFNs, we want to implement load balancing when forwarding traffic to upstream peers. For VFNs, we want all their mempool traffic to be forwarded to the connected validator.